### DISCUSSION TITLE
Not able to parse image information along with text and table in correct sequence

### DISCUSSION THREAD

**Discussion by reema93jain** [pymupdf/PyMuPDF#2982] [Q&A]
Reposting again as it seems like last discussion was closed:

Hi Team,

I am able to parse table and text data in order using below code. But now I also wanted to fetch image data along with text & table in sequence. When I tried adding image on below code, I am just getting repeated images and that not in alignment with text & table data on pdf.

Is there any logic which I can apply to below code to be able to fetch images  in sequence along with text and table data? Below is my code:

import fitz # PyMuPDF
from PIL import Image
from io import BytesIO
import matplotlib.pyplot as plt

def in_table(tab_bbox, line_bbox):
tab_rect = fitz.Rect(tab_bbox)
line_rect = fitz.Rect(line_bbox)
return line_rect.intersects(tab_rect)

def parse_pdf(pdf_path):
# Open pdf
doc = fitz.open(pdf_path)

#parsed_data=[]

# Iterate through all pages in the PDF document
for page_num in range(doc.page_count):
    page = doc[page_num]

    # Extract text
    lines = []  # will contain our lines outside any table

    # Find Tables
    tabs = page.find_tables()

    if tabs:
        for tab in tabs:
            #table=[]
            for line in tab.extract():
                lines.append(line)

    for block in page.get_text("dict", flags=fitz.TEXTFLAGS_TEXT)["blocks"]:  # look at text blocks only
        for line in block["lines"]:
            if any(in_table(tab.bbox, line["bbox"]) for tab in tabs):
                continue  # this line is in a table
            lines.append(line)
    #print(lines)
    #lines.sort(key=lambda l: (l["bbox"][3], l["bbox"][0]))  # sort lines in reading sequence
    
    #Checking if each line is list or dict. If its dict, print text if not, find tables as dataframe
    text2=""
    dlist=[]
    flag=0
    for line in lines:
        if isinstance(line, list):
            flag=flag+1
            if flag==1:
                columns = line
                #print(columns)
                flag=2
            else:
                dlist.append(line) 
                #print(dlist)
        if isinstance(line, dict) and "spans" in line:
            if dlist:
                    df = pd.DataFrame(data=dlist, columns=columns)
                    print(df)    
                    flag=0
                    dlist=[]
            for span in line['spans']:
                text = span["text"]
                #text2="".join(text)
                text2=text2+text+'\n'
    if dlist:
        df = pd.DataFrame(data=dlist, columns=columns)
        print(df)
        flag=0
        dlist=[]
    
    print(text2)
            
    
    # Extract images
    parsed_data=[]
    images = page.get_images(full=True)
    if images:
        for img_index, img in enumerate(images):
            img_bytes = doc.extract_image(img[0])
            image = Image.open(BytesIO(img_bytes["image"]))
            parsed_data.append({'type': 'image', 'content': image})
            
    for entry in parsed_data:
        if entry['type'] == 'image':
            image=entry['content']
           #print(entry['content'])
            plt.imshow(image)
            plt.show()
doc.close()     
Calling the function
pdf_path = "EOS-User-Manual.pdf"
parse_data = parse_pdf(pdf_path)
#calling sub-set
parsed_data=parse_data[0:100]
print(parsed_data)
print()

Thank you
Reema Jain

**Comment by reema93jain** [comment]
Hi Team,

I tried below code but it seems like I am not able to parse images properly along with text and table data. Below code is showing bunch of images together and missing all text/tables in between ,which is not correct as per pdf file.

Can someone please advise what else can be done here?

import fitz  # PyMuPDF
from PIL import Image
from io import BytesIO
import matplotlib.pyplot as plt
import pandas as pd


def in_table(tab_bbox,image_bbox,line_bbox):
    tab_rect = fitz.Rect(tab_bbox)
    line_rect = fitz.Rect(line_bbox)
    image_rect= fitz.Rect(image_bbox)
    return line_rect.intersects(tab_rect) and  line_rect.intersects(image_rect)

def parse_pdf(pdf_path):
    # Open pdf
    doc = fitz.open(pdf_path)
    
    
    # Iterate through all pages in the PDF document
    for page_num in range(doc.page_count):
        page = doc[page_num]

        # Extract text
        lines = []  # will contain our lines outside any table

        # Find Tables
        tabs = page.find_tables()

        if tabs:
            for tab in tabs:
                #table=[]
                for line in tab.extract():
                    lines.append(line)

        #Find Images                
        images = page.get_images(full=True)
        if images:
            for img_index, img in enumerate(images):
                img_bytes = doc.extract_image(img[0])
                image = Image.open(BytesIO(img_bytes["image"]))
                lines.append(image)  
        
        
        for block in page.get_text("dict", flags=fitz.TEXTFLAGS_TEXT)["blocks"]:  # look at text blocks only
            for line in block["lines"]:
                if any(in_table(tab.bbox,image.bbox,line["bbox"]) for tab in tabs):
                    continue  # this line is in a table & image                  
                lines.append(line)
        
     
        
        #print(lines)
        #lines.sort(key=lambda l: (l["bbox"][3], l["bbox"][0]))  # sort lines in reading sequence

        #print(f"Page {page_num + 1}:\n")
        #Checking if each line is list or dict. If its dict, print text. If its list,print tables as dataframe        
        text2=""
        dlist=[]
        flag=0
        for line in lines:
            if isinstance(line, list):
                flag=flag+1
                if flag==1:
                    columns = line
                    #print(columns)
                    flag=2
                else:
                    dlist.append(line) 
                    #print(dlist)
            if isinstance(line, dict) and "spans" in line:
                if dlist:
                    df = pd.DataFrame(data=dlist, columns=columns)
                    print(df)    
                    flag=0
                    dlist=[]
                for span in line['spans']:
                    text = span["text"]
                    #text2="".join(text)
                    text2=text2+text+'\n'
                print(text2)
                text2=""
            #if isinstance(line, Image.Image):
            else:
                if dlist:
                    df = pd.DataFrame(data=dlist, columns=columns)
                    print(df)    
                    flag=0
                    dlist=[]
                plt.imshow(line)
                plt.show()    
            
        if dlist:
            df = pd.DataFrame(data=dlist, columns=columns)
            print(df)
            flag=0
            dlist=[]
        
        #print(text2)        
                
    doc.close()     
        
  
        
# Calling the function
pdf_path = "EOS-User-Manual.pdf"
parse_data = parse_pdf(pdf_path)
#calling sub-set
parsed_data=parse_data[0:100]
print(parsed_data)
print()



Thank you
Reema Jain

**Comment by JorjMcKie** [comment]
You have not attached the PDF, so there is no way to compare your comments and code with the file's data.

Otherwise, I have problems to understand your code:
Are you appending image binary data to a list of lines which also contain text? Why? This is bound to produce unusable stuff.
